pkg/operator/status.go

```diff
@@ -124,45 +116,77 @@ func isDeploymentStatusComplete(o runtime.Object) bool {
 	return false
 }
 
-func (c *Controller) syncDeploymentStatus(cr *regopapi.ImageRegistry, o runtime.Object, statusChanged *bool) {
-	metaObject := o.(metaapi.Object)
+func (c *Controller) syncStatus(cr *regopapi.ImageRegistry, o runtime.Object, applyError error, statusChanged *bool) {
+	metaObject, _ := o.(metaapi.Object)
 
 	operatorAvailable := osapi.ConditionFalse
 	operatorAvailableMsg := ""
-
-	if metaObject.GetDeletionTimestamp() == nil && isDeploymentStatusAvailable(o) {
+	if o == nil {
+		operatorAvailableMsg = "deployment does not exist"
+	} else if metaObject.GetDeletionTimestamp() != nil {
+		operatorAvailableMsg = "deployment is being deleted"
+	} else if !isDeploymentStatusAvailable(o) {
+		operatorAvailableMsg = "deployment does not have available replicas"
+	} else {
 		operatorAvailable = osapi.ConditionTrue
 		operatorAvailableMsg = "deployment has minimum availability"
 	}
 
-	errOp := c.clusterStatus.Update(osapi.OperatorAvailable, operatorAvailable, operatorAvailableMsg)
-	if errOp != nil {
-		logrus.Errorf("unable to update cluster status to %s=%s: %s", osapi.OperatorAvailable, osapi.ConditionTrue, errOp)
+	err := c.clusterStatus.Update(osapi.OperatorAvailable, operatorAvailable, operatorAvailableMsg)
+	if err != nil {
+		logrus.Errorf("unable to update cluster status to %s=%s (%s): %s", osapi.OperatorAvailable, operatorAvailable, operatorAvailableMsg, err)
 	}
 
-	operatorProgressing := osapi.ConditionTrue
-	operatorProgressingMsg := "deployment is progressing"
+	updateCondition(cr, &operatorapi.OperatorCondition{
+		Type:               operatorapi.OperatorStatusTypeAvailable,
+		Status:             operatorapi.ConditionStatus(operatorAvailable),
+		LastTransitionTime: metaapi.Now(),
+		Message:            operatorAvailableMsg,
+	}, statusChanged)
 
-	if metaObject.GetDeletionTimestamp() == nil {
-		if isDeploymentStatusComplete(o) {
-			operatorProgressing = osapi.ConditionFalse
-			operatorProgressingMsg = "deployment successfully progressed"
-		}
+	operatorProgressing := osapi.ConditionTrue
+	operatorProgressingMsg := ""
+	if applyError != nil {
+		operatorProgressingMsg = fmt.Sprintf("unable to apply resources: %s", applyError)
+	} else if o == nil {
+		operatorProgressingMsg = "all resources are successfully applied, but the deployment does not exist"
+	} else if metaObject.GetDeletionTimestamp() != nil {
+		operatorProgressingMsg = "the deployment is being deleted"
+	} else if !isDeploymentStatusComplete(o) {
+		operatorProgressingMsg = "the deployment has not completed"
 	} else {
 		operatorProgressing = osapi.ConditionFalse
-		operatorProgressingMsg = "deployment removed"
+		operatorProgressingMsg = "everything is ready"
 	}
```

Comment:
Progressing needs to be false if the current active deployment does not match the operator configuration.

So for example if the AWS credential secret has been updated, and the current deployment has not mounted that secret, progressing should be true.  (We would detect this by determining that the credential hash annotated on the deployment does not match the hash of the current credentials in the credential secret).

Similarly if the registry resource itself has been updated, but the current deployment doesn't reflect those changes...so again, hash of the registry resource, annotated onto the registry deployment, and compare them.

Reply:
Did you want to say "Progressing needs to be *true* if the current active deployment does not match the operator configuration"? I think progressing should be true if we haven't reached the desired state. And that's what was implemented here.

Reply:
Sorry, yes that is what i meant to say.

I think all the existing logic here is fine, but i think it's missing the case i described where the current deployment is complete(all replicas active/rolled out), but the current deployment doesn't reflect the desired configuration.  Unless "isDeploymentStatusComplete" is actually checking that?

Reply:
This function is called after all resources are updated according to the custom resource, so if applyError is nil, the deployment should reflect the desired configuration. There might be a race if someone touched the deployment just after we had updated it. But in this case we would receive an event for the deployment and run this function again.

Reply:
ok